File: 2025/1/week2/예산_백준_2512/Solution.java
import java.io.*;
import java.util.*;

class Solution{
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int n = Integer.parseInt(br.readLine());
        int[] budgetReqs = new int[n];

        int lowBound = 1;
        int highBound = 0;
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            budgetReqs[i] = Integer.parseInt(st.nextToken());
            highBound = Math.max(highBound, budgetReqs[i]);
        }
        int totalBudget = Integer.parseInt(br.readLine());

        int result = lowBound;
        int midBound ;
        while(lowBound <= highBound){
            midBound = (lowBound + highBound)/2;
            int sumBudgetReq = 0;
            for (int budgetReq : budgetReqs) {
                sumBudgetReq += Math.min(budgetReq, midBound);
            }
            if (sumBudgetReq <= totalBudget){
                result = midBound;
                lowBound = midBound + 1;
            }else{
                highBound = midBound -1;
            }
        }
        System.out.println(result);
    }
}

File: 2025/1/week2/예산_백준_2512/Solution.py
import sys

input = sys.stdin.readline

n = int(input())
budgetReqs = list(map(int, input().split()))
totalBudget = int(input())

minBoundary = 1
maxBoundary = max(budgetReqs)
midBoundary = (minBoundary + maxBoundary) // 2
result = midBoundary
while minBoundary <= maxBoundary:
    midBoundary = (minBoundary + maxBoundary) // 2
    sumBudgetReq = 0
    for budgetReq in budgetReqs:
        sumBudgetReq += min(budgetReq, midBoundary)
    if sumBudgetReq <= totalBudget:
        minBoundary = midBoundary +1
        result = midBoundary
    else:
        maxBoundary=midBoundary-1
print(result)